includes/class-newspack-listings-taxonomies.php

```diff
@@ -0,0 +1,664 @@
+		$shadow_taxonomies = self::get_shadow_taxonomy_config();
+
+		// Register shadow taxonomies for each source post type.
+		foreach ( $shadow_taxonomies as $post_type_to_shadow => $shadow_taxonomy ) {
+			$taxonomy_slug = self::NEWSPACK_LISTINGS_TAXONOMIES[ $post_type_to_shadow ];
+			register_taxonomy(
+				$taxonomy_slug,
+				$shadow_taxonomy['post_types'],
+				$shadow_taxonomy['config']
+			);
+		}
+	}
+
+	/**
+	 * Using the array of shadow taxonomies, get an array of corresponding post types being shadowed.
+	 *
+	 * @return array Array of post types being shadowed.
+	 */
+	public static function get_post_types_to_shadow() {
+		return array_values(
+			array_intersect_key(
+				Core::NEWSPACK_LISTINGS_POST_TYPES,
+				self::NEWSPACK_LISTINGS_TAXONOMIES
+			)
+		);
+	}
+
+	/**
+	 * Given a post type, get the name of the shadow taxonomy for that post type.
+	 *
+	 * @param string $post_type Post type.
+	 * @return string}boolean The shadow taxonomy for the given post type, or false if none.
+	 */
+	public static function get_taxonomy_by_post_type( $post_type ) {
+		$shadow_taxonomy = false;
+		$post_type_slugs = array_keys( Core::NEWSPACK_LISTINGS_POST_TYPES, $post_type );
+		$post_type_slug  = reset( $post_type_slugs );
+
+		if ( ! empty( $post_type_slug ) && ! empty( self::NEWSPACK_LISTINGS_TAXONOMIES[ $post_type_slug ] ) ) {
+			$shadow_taxonomy = self::NEWSPACK_LISTINGS_TAXONOMIES[ $post_type_slug ];
+		}
+
+		return $shadow_taxonomy;
+	}
+
+	/**
+	 * Create shadow relationships between taxonomies and their posts.
+	 */
+	public static function create_shadow_relationship() {
+		add_action( 'wp_insert_post', [ __CLASS__, 'update_or_delete_shadow_term' ], 10, 2 );
+		add_action( 'before_delete_post', [ __CLASS__, 'delete_shadow_term' ] );
+	}
+
+	/**
+	 * When a listing post type changes status, add/update its shadow term if the status is `publish`, otherwise delete it.
+	 *
+	 * @param int   $post_id ID for the post being inserted or saved.
+	 * @param array $post Post object for the post being inserted or saved.
+	 * @return void
+	 */
+	public static function update_or_delete_shadow_term( $post_id, $post ) {
+		// Get the taxonomy to update or delete.
+		$shadow_taxonomy = self::get_taxonomy_by_post_type( $post->post_type );
+
+		// Bail if not a shadowable post.
+		if ( ! $shadow_taxonomy ) {
+			return;
+		}
+
+		// If the post is a valid post, update or create the shadow term. Otherwise, delete it.
+		if ( self::validate_post( $post ) ) {
+			self::update_shadow_term( $post, $shadow_taxonomy );
+		} else {
+			self::delete_shadow_term( $post, $shadow_taxonomy );
+		}
+	}
+
+	/**
+	 * Validate a post object to check whether it should have a shadow term.
+	 *
+	 * @param object $post Post object to validate.
+	 * @return bool True if the post should have a shadow term, otherwise false.
+	 */
+	public static function validate_post( $post ) {
+		$is_valid_post = true;
+
+		// Post must be published.
```

Comment:
```suggestion
		// Post must not be published.
```
Since the check below was updated. But that's weird, because I'd expect a published post to be linkable to a listing. Is that check below intended?

Reply:
It's somewhat confusing (I even confused myself!) because `validate_post` returns `true` if the listing is "valid" (e.g. it should have a shadow term), and `false` if it's not valid. So the check should be: if the post is **not** published, return `false`, which is what the change in 4b43a7c reflects.

Reply:
Then the word "validate" is perhaps not the best - how about `should_update_shadowterm` as the method name?
Also, the code comment should still be updated ;)

Reply:
`should_update_shadow_term` does make the function's purpose clearer. Updated in a136170.